### ISSUE TITLE
KeyboardInterrupt is ignored when subscribing

### ISSUE THREAD

**Issue by galah92** [googleapis/python-pubsub#262]
#### Environment details

  - OS type and version: Windows 10
  - Python version: Python 3.8.6
  - pip version: pip 20.3.3
  - `google-cloud-pubsub` version: 2.2.0

#### Code example

```python
def callback(message):
    print(message.data.decode('utf-8'))
    message.ack()


def run():
    subscriber = pubsub_v1.SubscriberClient()
    future = subscriber.subscribe(SUBSCRIPTION_PATH, callback=callback)
    try:
        future.result()
    except KeyboardInterrupt:
        future.cancel()


if __name__ == '__main__':
    run()

```

The following code seems to hang and ignore `CTRL-C`. Messages are keep on coming. This doesn't happen on Linux (tried on WSL).

Thanks

**Comment by plamut** [comment]
I tried reproducing this on my laptop (Ubuntu 18.04) and the keyboard interrupt worked fine when running the sample from the command line. However, when running the very same sample using the very same Python 3.8 interpreter through the IDE (VS Code in my case), the keyboard interrupt was **not** caught, regardless of whether I ran it it debug mode or not.

I also tried running `sleep 10` from the IDE's built-in terminal and was able to interrupt it with <kbd>Ctrl</kbd>+<kbd>C</kbd>, meaning that the IDE's terminal is probably not the culprit.

I vaguely remember a similar issue from the past with `grpcio` version `1.22.0` (I think), and upgrading the dependency worked then, but not in this case (tested with the latest `grpcio==1.34.1`).

**Comment by galah92** [comment]
@plamut Thanks.
I'll add that this also reproducible when running the script on Windows Terminal, so this has nothing to do with VSCode.

**Comment by lidizheng** [comment]
Just add context here, @plamut maybe referring to this issue https://github.com/grpc/grpc/issues/19973, that ctrl-c might cause deadlock with a small chance, and it's fixed in `1.24.0`.

Windows' ctrl-c is different than Linux. Can you try to register a [signal handler](https://docs.python.org/3/library/signal.html) for the application to see if the signal is received by the process?

**Comment by plamut** [comment]
@lidizheng I tested on Ubuntu Linux yet again and varied the following:
 - Running the script in VS Code or in the plain terminal,
 - SIGINT handler registered (yes/no)
 - Fired <kbd>Ctrl</kbd> + <kbd>C</kbd> when blocked by `future.result()` or by a `time.sleep(5)` call that I added.
 
 <table style="border-collapse: collapse; width: 100%; height: 90px;" border="1">
<tbody>
<tr style="height: 18px;">
<td style="width: 16.6666%; height: 18px;">&nbsp;</td>
<td style="width: 16.6666%; height: 18px;">&nbsp;</td>
<td style="width: 33.3333%; height: 18px;"><strong>Interrupt time.sleep()</strong></td>
<td style="width: 33.3333%; height: 18px;"><strong>interrupt future.result()</strong></td>
</tr>
<tr style="height: 18px;">
<td style="width: 16.6666%; height: 36px;" rowspan="2"><strong>Plain&nbsp; terminal</strong></td>
<td style="width: 16.6666%; height: 18px;"><strong>SIGINT handler (no)</strong></td>
<td style="width: 33.3333%; height: 18px;">YES</td>
<td style="width: 33.3333%; height: 18px;">YES</td>
</tr>
<tr style="height: 18px;">
<td style="width: 16.6666%; height: 18px;"><strong>SIGINT handler (yes)</strong></td>
<td style="width: 33.3333%; height: 18px;">YES</td>
<td style="width: 33.3333%; height: 18px;">YES</td>
</tr>
<tr style="height: 18px;">
<td style="width: 16.6666%; height: 36px;" rowspan="2"><strong>VS Code</strong></td>
<td style="width: 16.6666%; height: 18px;"><strong>SIGINT handler (no)</strong></td>
<td style="width: 33.3333%; height: 18px;">NO</td>
<td style="width: 33.3333%; height: 18px;">NO</td>
</tr>
<tr style="height: 18px;">
<td style="width: 16.6666%; height: 18px;"><strong>SIGINT handler (yes)</strong></td>
<td style="width: 33.3333%; height: 18px;">NO</td>
<td style="width: 33.3333%; height: 18px;">YES</td>
</tr>
</tbody>
</table>

Things work fine in the terminal, but apparently VS Code somewhat interferes with signals.
(my previous `sleep 10` test was a bash command ran the VS Code's integrated terminal directly, i.e. without starting an actual Python script)

All in all things seem to work just fine on Linux. Any "lost" interrupts can be attributed to VS Code and not the library code itself.

It might be different on Windows, but I don't have a Windows machine at hand right now to test (not even virtual).

**Comment by lidizheng** [comment]
@galah92 Can you help to test your environment with signal handler to see if the Python process received the signal? Or better, use a debugger (usually IDEs will contain one), it will be very explicit about what signal is received, and where does it freeze.

(I also don't have access to a Windows machine)

**Comment by galah92** [comment]
Hi, sorry for the late response. I tested it today again. WSL2 (using Windows Terminal) works. Windows doesn't work no matter what I tried: VSCode, VSCode using a debugger (didnt catch the signal), running from Windows Terminal or straight from the Command Line. I'm not sure how to debug this exactly.

**Comment by plamut** [comment]
@galah92 Thanks for the info!

It seems that the signal gets lost at a level below Python, most likely in the underlying gRPC dependency. As such I will mark it as external and wait on more info from @lidizheng when he finds a time slot to have a look at this.

**Comment by lidizheng** [comment]
@galah92 Thanks for the investigation. Since you have a debugger environment `VSCode using a debugger`, do you think you can validate if other signals are being processed, (like SIGTERM, SIGABRT)? In the past, gRPC Python did swallow signals because the signal arrived when the process is running in C extension space. We fixed it by checking [`PyErr_CheckSignals `](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi#L42) explicitly years ago.

If the SIGTERM works, can you share the stacktrace (most importantly, the C/C++ stacktraces)? So, we can know when the SIGINT is ignored, what is occupying the gRPC library.

**Comment by acocuzzo** [comment]
@galah92 are you still experiencing this issue?

**Comment by galah92** [comment]
I moved from Windows to MacOS / Linux machines, don't have a Windows machine anymore to test it with, sorry.

**Comment by Mithil467** [comment]
@acocuzzo I am not sure how to send these signals on windows or if it even is possible. Are you aware of any process that I can follow?

**Comment by acocuzzo** [comment]
@Mithil467 Can you try to follow what @lidizheng suggested, to register a [signal handler](https://docs.python.org/3/library/signal.html) for the application to see if the signal is received by the process?
Thanks, sorry I don't have access to a Windows machine to test.

**Comment by Mithil467** [comment]
I added the handler inside script -
```py
import signal
def handler(signum, frame):
    print("signal handler called with signal", signum)
signal.signal(signal.SIGINT, handler)
signal.signal(signal.SIGBREAK, handler)
```
Then while the future.result(timeout=timeout) method was running, I pressed CTRL-C. It didn't stop the program immediately but waited for the timeout to expire. Then printed - `signal handler called with signal 2`. I was also able to send SIGBREAK, and observed same behaviour.

I tried this on WSL and there the signal works immediately when CTRL-C is pressed.

It actually might just be the way in which Windows and Linux deal with threads and signals, in which case we might not have a solution to this.
